### ISSUE TITLE
parallel build fails

### ISSUE THREAD

**Issue by Whissi** [numpy/numpy#13080]
Hi,

on Gentoo Linux, we see parallel build issues in all versions (including latest 1.16.x version) for Python 3.x:

Most common failure with lapack (should come close to your defaults):

[build-error_py36_with_lapack.txt](https://github.com/numpy/numpy/files/2923505/build-error_py36_with_lapack.txt)

However, this is not always reproducible.

Most reproducible failure is without lapack:

[build-error_py36_without_lapack.txt](https://github.com/numpy/numpy/files/2923506/build-error_py36_without_lapack.txt)

Downstream-Bug: https://bugs.gentoo.org/660754

**Comment by charris** [comment]
What is the Gentoo build infrastructure? I take it that this is from source.

**Comment by rgommers** [comment]
Hmm, not entirely clear what is going on. Some warnings about append behavior. Are `FFLAGS` and/or `LDFLAGS` set, and if so to what? And to be sure, it's *only* the parallel builds that are failing?

```
/var/tmp/portage/dev-python/numpy-1.16.1/work/numpy-1.16.1/numpy/distutils/fcompiler/__init__.py:517: UserWarning: FFLAGS is used as is, not appended to flags already defined by numpy.distutils! Use NPY_DISTUTILS_APPEND_FLAGS=1 to obtain appending behavior instead (this behavior will become default in a future release).
  fflags = self.flag_vars.flags + dflags + oflags + aflags
/var/tmp/portage/dev-python/numpy-1.16.1/work/numpy-1.16.1/numpy/distutils/fcompiler/__init__.py:530: UserWarning: LDFLAGS is used as is, not appended to flags already defined by numpy.distutils! Use NPY_DISTUTILS_APPEND_FLAGS=1 to obtain appending behavior instead (this behavior will become default in a future release).
  linker_so_flags = self.flag_vars.linker_so
/var/tmp/portage/dev-python/numpy-1.16.1/work/numpy-1.16.1/numpy/distutils/fcompiler/__init__.py:540: UserWarning: LDFLAGS is used as is, not appended to flags already defined by numpy.distutils! Use NPY_DISTUTILS_APPEND_FLAGS=1 to obtain appending behavior instead (this behavior will become default in a future release).
  linker_exe_flags = self.flag_vars.linker_exe
error: Command "x86_64-pc-linux-gnu-gcc -O2 -pipe -march=znver1 -mtune=znver1 -shared -Wl,-O1 -Wl,--as-needed -shared -O2 -pipe -march=znver1 -mtune=znver1 -fno-strict-aliasing /var/tmp/portage/dev-python/numpy-1.16.1/work/numpy-1.16.1-python3_6/temp.linux-x86_64-3.6/var/tmp/portage/dev-python/numpy-1.16.1/work/numpy-1.16.1-python3_6/src.linux-x86_64-3.6/numpy/core/src/multiarray/_multiarray_tests.o /var/tmp/portage/dev-python/numpy-1.16.1/work/numpy-1.16.1-python3_6/temp.linux-x86_64-3.6/numpy/core/src/common/mem_overlap.o -L/usr/lib64 -L/var/tmp/portage/dev-python/numpy-1.16.1/work/numpy-1.16.1-python3_6/temp.linux-x86_64-3.6 -lnpymath -lpython3.6m -o /var/tmp/portage/dev-python/numpy-1.16.1/work/numpy-1.16.1-python3_6/lib/numpy/core/_multiarray_tests.cpython-36m-x86_64-linux-gnu.so" failed with exit status 1
```

**Comment by Whissi** [comment]
> What is the Gentoo build infrastructure? I take it that this is from source.

Yes, we don't have a centralized build infrastructure. Every Gentoo user is building from source. In my case I provided build.log excerpts from systems with the following specs (of course, different processors but toolchain is identical):
```
# emerge --info
Portage 2.3.62 (python 3.6.6-final-0, default/linux/amd64/17.1, gcc-8.3.0, glibc-2.28-r5, 4.20.13-gentoo x86_64)
=================================================================
System uname: Linux-4.20.13-gentoo-x86_64-Intel-R-_Core-TM-_i7-3770K_CPU_@_3.50GHz-with-gentoo-2.6
KiB Mem:     8134048 total,   3543768 free
KiB Swap:    1048572 total,   1048572 free
Timestamp of repository gentoo: Sun, 03 Mar 2019 13:44:13 +0000
Head commit of repository gentoo: 11fbab6c642975873b51f9b6b4b7aa43a3a7ec4e

sh bash 5.0_p2
ld GNU ld (Gentoo 2.32 p1) 2.32.0
distcc 3.3.2 x86_64-pc-linux-gnu [disabled]                                                                             ccache version 3.6 [disabled]                                                                                           app-shells/bash:          5.0_p2::gentoo
dev-java/java-config:     2.2.0-r4::gentoo
dev-lang/perl:            5.28.0::gentoo
dev-lang/python:          2.7.15::gentoo, 3.6.6::gentoo, 3.7.0::gentoo
dev-util/ccache:          3.6::gentoo
dev-util/cmake:           3.13.4::gentoo
dev-util/pkgconfig:       0.29.2::gentoo
sys-apps/baselayout:      2.6-r1::gentoo
sys-apps/openrc:          0.41.2::gentoo
sys-apps/sandbox:         2.15::gentoo
sys-devel/autoconf:       2.13-r1::gentoo, 2.69-r4::gentoo
sys-devel/automake:       1.16.1-r1::gentoo
sys-devel/binutils:       2.31.1-r4::gentoo, 2.32::gentoo
sys-devel/gcc:            7.3.0-r6::gentoo, 8.2.0-r6::gentoo, 8.3.0::gentoo
sys-devel/gcc-config:     2.0::gentoo
sys-devel/libtool:        2.4.6-r5::gentoo
sys-devel/make:           4.2.1-r4::gentoo
sys-kernel/linux-headers: 4.20::gentoo (virtual/os-headers)
sys-libs/glibc:           2.28-r5::gentoo
```

@rgommers If build system will honor system flags, `FFLAGS` initial value should be set to "-O2 -pipe" and `LDFLAGS` initial value should be set to "-Wl,-O1 -Wl,--as-needed". We do some changes later, see https://gitweb.gentoo.org/repo/gentoo.git/tree/dev-python/numpy/numpy-1.16.1.ebuild?id=d5337dedc68adc4dfcbfe8bb045c96554512aabf

Please see a build.log excerpts from a successful build using `MAKEOPTS=-j1`:

[build-success_py36_without_lapack_MAKEOPTS-j1.txt](https://github.com/numpy/numpy/files/2923651/build-success_py36_without_lapack_MAKEOPTS-j1.txt)

**Comment by rgommers** [comment]
> @rgommers If build system will honor system flags, FFLAGS initial value should be set to "-O2 -pipe" and LDFLAGS initial value should be set to "-Wl,-O1 -Wl,--as-needed"

yes and no - we do use `FFLAGS` and `LDFLAGS`, but the default behavior is to use those as as rather than append them to the defaults. It's unclear from your link to that patch if that's taken into account (probably yes though, since that's how it has always worked). If you want to use append behavior (which is what most users expect), then you need to set `NPY_DISTUTILS_APPEND_FLAGS=1`.

**Comment by Whissi** [comment]
Now I think I understand. OK, we will set `NPY_DISTUTILS_APPEND_FLAGS=1` in future but it has no effect on the current problem:

[build-error_py36_without_lapack-with-NPY_DISTUTILS_APPEND_FLAGS.txt](https://github.com/numpy/numpy/files/2923679/build-error_py36_without_lapack-with-NPY_DISTUTILS_APPEND_FLAGS.txt)

**Comment by rgommers** [comment]
Hmm, that last build log still doesn't contain `-g` and `-Wall` which should be there (they're gfortran `linker_so` flags defined in https://github.com/numpy/numpy/blob/master/numpy/distutils/fcompiler/gnu.py#L97. So not sure if the append behavior was changed there.

```
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=znver1 -mtune=znver1 -shared -Wl,-O1 -Wl,--as-needed -shared -O2 -pipe -march=znver1 -mtune=znver1 -fno-strict-aliasing /var/tmp/portage/dev-python/numpy-1.16.1/work/numpy-1.16.1-python3_6/temp.linux-x86_64-3.6/var/tmp/portage/dev-python/numpy-1.16.1/work/numpy-1.16.1-python3_6/src.linux-x86_64-3.6/numpy/core/src/multiarray/_multiarray_tests.o /var/tmp/portage/dev-python/numpy-1.16.1/work/numpy-1.16.1-python3_6/temp.linux-x86_64-3.6/numpy/core/src/common/mem_overlap.o -L/usr/lib64 -L/var/tmp/portage/dev-python/numpy-1.16.1/work/numpy-1.16.1-python3_6/temp.linux-x86_64-3.6 -lnpymath -lpython3.6m -o /var/tmp/portage/dev-python/numpy-1.16.1/work/numpy-1.16.1-python3_6/lib/numpy/core/_multiarray_tests.cpython-36m-x86_64-linux-gnu.so
/usr/lib/gcc/x86_64-pc-linux-gnu/8.3.0/../../../../x86_64-pc-linux-gnu/bin/ld: /var/tmp/portage/dev-python/numpy-1.16.1/work/numpy-1.16.1-python3_6/temp.linux-x86_64-3.6/numpy/core/src/common/mem_overlap.o: file not recognized: file truncated
collect2: error: ld returned 1 exit statu
```
searching for this error, it's often caused by left-over object files from previous failed builds. Are you sure it's a clean rebuild?

**Comment by rgommers** [comment]
not sure I'm on the right track there though, because the compile flags should not depend on whether parallel build is enabled. but still, it looks wrong

**Comment by charris** [comment]
I was wondering what Portage was doing, how it implements parallel builds, sets paths, etc.

**Comment by Whissi** [comment]
Don't focus too much on Gentoo. I can reproduce the same on latest Debian Stretch:

```
mkdir /tmp/numpy-test
cd /tmp/numpy-test
wget https://files.pythonhosted.org/packages/cf/8d/6345b4f32b37945fedc1e027e83970005fc9c699068d2f566b82826515f2/numpy-1.16.2.zip
apt-get build-dep python-numpy
export {ATLAS,PTATLAS,BLAS,LAPACK,MKL}=None
export NPY_DISTUTILS_APPEND_FLAGS=1
export CC="x86_64-linux-gnu-gcc -O2 -pipe -march=native"
export FFLAGS="-O2 -pipe -fno-strict-aliasing -fPIC"
export LDFLAGS="-Wl,-O1 -Wl,--as-needed -shared"
unzip numpy-1.16.2.zip
cd numpy-1.16.2/
python3 setup.py build -j 31
```


[build-error_py35_Debian.txt](https://github.com/numpy/numpy/files/2923822/build-error_py35_Debian.txt)

**Comment by Whissi** [comment]
`export CC=...` seems to be enough. You don't even need to mess with FFLAGS or LDFLAGS.

**Comment by charris** [comment]
What about smaller number of builds? I run with `-j 4' all the time without problem, at leas installing from the repo.

**Comment by rgommers** [comment]
Okay I can reproduce this now, but it took a while. I only can reproduce it on current master for:

- building with `lapack_lite`
- using `export CC/LDFLAGS/FFLAGS`
- parallel build

**Comment by rgommers** [comment]
This is probably a duplicate of https://github.com/scipy/scipy/issues/7112#issuecomment-369993514. That comment has an explanation (handwaving, `distutils` parallelism bug rather than `numpy.distutils`).

**Comment by rth** [comment]
> This is probably a duplicate of scipy/scipy#7112 (comment). That comment has an explanation (handwaving, distutils parallelism bug rather than numpy.distutils).

It would be nice if someone could report it upstream to CPython. The only relevant issue I found is https://bugs.python.org/issue5309 which says that parallel compilation of extensions should have been fixed in Python 3.5.

Also, what's the long term vision of `numpy.distutils`? Are relevant changes expected to get merged into stdlib distutils at some point or have the code bases diverged to the point that it's unlikely to happen?

**Comment by rgommers** [comment]
> Also, what's the long term vision for numpy.distutils? Are relevant changes expected to get merged into stdlib distutils at some point or have the code bases diverged to the point that it's unlikely to happen?

@rth no that won't happen, for one because it's a lot of work and also because of reasons like CPython not caring about Fortran and we not being able to live with the slow CPython release cycles.

The plan is to remove `distutils` from the stdlib and merge it into `setuptools`. Then we need to adjust `numpy.distutils` for that.

**Comment by usiems** [comment]
Just my observation, because I seem to just have stumbled across the same issue:
For our group it always is a missing lapack symbol in the _umath_linalg module, which vanishes after a recompile (which is a fresh re-compile with Jenkins in a docker container, so the starting conditions should always be the same).

After comparing the logs I came to the conclusion that this must be an issue of order/timing. I also noticed that the _umath_linalg library is smaller in the error case. And a look at the setup.py of linalg revealed that there are two extensions built, _umath_linalg and lapack_lite (the latter only seems to be used for testing by the way), with almost the same lapack_lite source files, and the same source files are really built twice, at the same location. So I guess sometimes the object files are overwritten by the second compile while the linking for the first extension takes place, and that is the cause of our troubles.

An updated distutils should probably take these cases into account and, e.g., build each source file only once, or into different object files if the compile flags are different.

**Comment by rth** [comment]
As far as I understand, the future is to build with meson (https://github.com/numpy/numpy/pull/22663) which does build in parallel by default and shouldn't have such issues.

So maybe this issue could be closed, as distutils is being deprecated anyway.

**Comment by rgommers** [comment]
We'll still accept a PR with a fix if anyone has or needs it, but indeed we won't invest maintainer time in chasing this down. The Meson build support is already present in `main` (not 100% complete, but it works and all tests pass in CI across OSes). So yes, we'll close this once support for a distutils-based build is removed.

**Comment by mattip** [comment]
Closing since fixing distutils parallel compilation is a non-goal.
